File: backend-services/jwt_auth/src/jwks.rs
use serde::Deserialize;
use reqwest::Client;
use jsonwebtoken::DecodingKey;

#[derive(Debug, Deserialize, Clone)]
pub struct JwkKey {
    pub kty: String,
    pub kid: String,
    #[serde(rename = "use")]
    pub use_: String,
    pub alg: String,
    pub n: String,
    pub e: String,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Jwks {
    pub keys: Vec<JwkKey>,
}

impl Jwks {
    pub async fn fetch(authority: &str) -> anyhow::Result<Self> {
        let url = format!("{}/.well-known/jwks.json", authority);
        let jwks = Client::new().get(&url).send().await?.json::<Jwks>().await?;
        Ok(jwks)
    }

    pub fn get_decoding_key(&self, kid: &str) -> Option<Result<DecodingKey, jsonwebtoken::errors::Error>> {
        self.keys
            .iter()
            .find(|k| k.kid == kid)
            .map(|jwk| DecodingKey::from_rsa_components(&jwk.n, &jwk.e))
    }
}

File: game-server/pkg/auth/jwks.go
package auth

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type JwkKey struct {
	Kty string `json:"kty"`
	Kid string `json:"kid"`
	Use string `json:"use"`
	N   string `json:"n"`
	E   string `json:"e"`
	Alg string `json:"alg"`
}

type Jwks struct {
	Keys []JwkKey `json:"keys"`
}

func FetchJwks(authority string) (*Jwks, error) {
	url := fmt.Sprintf("%s/.well-known/auth.json", authority)

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}

	defer resp.Body.Close()

	var jwks Jwks
	if err := json.NewDecoder(resp.Body).Decode(&jwks); err != nil {
		return nil, err
	}

	return &jwks, nil
}
